### PR TITLE
Enable the pod garbage collector

### PR DESCRIPTION
**Pull request by tiwillia**
Re-implented this, it is still failing for the same reason as the previous PR:

  https://github.com/openshift/origin/pull/5540

### CODE REVIEW

**Comment by tiwillia** [comment]
FYI @dgoodwin

**Comment by tiwillia** [comment]
As an example, creating a project (testproject1, in this example) and deleting the project results in the following:
```
namespace_controller_utils.go:208] Syncing namespace testproject1
tokens_controller.go:189] Deleting secret testproject1/builder-token-duaxi because service account builder was deleted
tokens_controller.go:189] Deleting secret testproject1/builder-token-wpmyu because service account builder was deleted
tokens_controller.go:189] Deleting secret testproject1/default-token-2jx79 because service account default was deleted
tokens_controller.go:189] Deleting secret testproject1/default-token-gr8xy because service account default was deleted
tokens_controller.go:189] Deleting secret testproject1/deployer-token-zyhdn because service account deployer was deleted
tokens_controller.go:189] Deleting secret testproject1/deployer-token-sbjz3 because service account deployer was deleted
tokens_controller.go:191] Error deleting secret testproject1/deployer-token-sbjz3: secrets "deployer-token-sbjz3" not found
namespace_controller.go:142] Finished syncing namespace "testproject1" (496ns)
namespace_controller_utils.go:208] Syncing namespace testproject1
namespace_controller.go:142] Finished syncing namespace "testproject1" (433ns)
namespace_controller_utils.go:208] Syncing namespace testproject1
namespace_controller.go:142] Finished syncing namespace "testproject1" (420ns)
namespace_controller_utils.go:208] Syncing namespace testproject1
namespace_controller.go:142] Finished syncing namespace "testproject1" (361ns)
namespace_controller_utils.go:208] Syncing namespace testproject1
namespace_controller.go:142] Finished syncing namespace "testproject1" (432ns)
namespace_controller_utils.go:208] Syncing namespace testproject1
```

`Syncing namespace` and `Finished syncing` repeat eternally.

File: pkg/cmd/server/start/start_master.go [hunk 1]
```diff
@@ -586,6 +586,7 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 		kc.RunPersistentVolumeClaimBinder(binderClient)
 		kc.RunPersistentVolumeProvisioner(provisionerClient)
 		kc.RunPersistentVolumeClaimRecycler(oc.ImageFor("recycler"), recyclerClient, oc.Options.PolicyConfig.OpenShiftInfrastructureNamespace)
+		kc.RunGCController()
```
Comment (ironcladlou) [inline pkg/cmd/server/start/start_master.go]: `go kc.RunGCController()` otherwise this call blocks.

Comment (liggitt) [inline pkg/cmd/server/start/start_master.go]: or move the goroutine spawn into the RunGCController func to match the others

Comment (tiwillia) [inline pkg/cmd/server/start/start_master.go]: Modified `RunGCController` to match others by starting the controller in a goroutine. Issue should be resolved now.

**Comment by tiwillia** [comment]
@derekwaynecarr could you review this when you have time?

**Comment by pweil-** [comment]
@deads2k @liggitt

File: pkg/cmd/server/kubernetes/master.go [hunk 2]
```diff
@@ -242,6 +243,13 @@ func (c *MasterConfig) RunResourceQuotaManager() {
 	go resourceQuotaManager.Run(c.ControllerManager.ConcurrentResourceQuotaSyncs, utilwait.NeverStop)
 }
 
+func (c *MasterConfig) RunGCController() {
+	if c.ControllerManager.TerminatedPodGCThreshold > 0 {
+		gcController := gccontroller.New(internalclientset.FromUnversionedClient(c.KubeClient), kctrlmgr.ResyncPeriod(c.ControllerManager), c.ControllerManager.TerminatedPodGCThreshold)
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master.go]: Take a look at something like the persistent volume controllers that make use of service account credentials to run the controller.  That's the model we want to follow for controllers as we move forward.  Limited permissions.

**Comment by tiwillia** [comment]
@deads2k I /think/ I've made the changes you recommended. Can you check this out again?

**Comment by deads2k** [comment]
> @deads2k I /think/ I've made the changes you recommended. Can you check this out again?

Your unit tests will fail.  Also, you should locate the kube e2e tests and run them here.  @smarterclayton do we always want to exclude `[Slow]` from `core.sh`?  The first GC one I see is slow.

**Comment by smarterclayton** [comment]
We always want to exclude [Slow].  But you can have it specifically run that test in a third step (after manual and the serial ones).

**Comment by tiwillia** [comment]
Unit tests should be passing now [test]

@deads2k is there some integration with github to run the kube e2e tests here? I have them running externally and can ensure they are all passing, if that is what you meant.

**Comment by deads2k** [comment]
@soltysh help me with the string here.  I think its:
[test][extended:core(Garbage collector)]
to run https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/test/e2e/garbage_collector.go#L32 .

@tiwillia can you make a `test/extended/slow.sh` that only runs this test for now, but is structured for us to easily add them?  Then we'll want a jenkins job and you should sign up for failure notifications.

**Comment by soltysh** [comment]
@deads2k you nailed it, for reference have a look [here](https://github.com/openshift/vagrant-openshift/blob/4008be434aad1a719d31bfb581921affe01a7ace/lib/vagrant-openshift/action/run_origin_tests.rb#L147-L149).

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
test https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2025/ didn't try to run it.  @soltysh can you try to help get it test.

@tiwillia you should probably try to run it locally.

**Comment by deads2k** [comment]
[test][extended:core(collector)]

just in case my bad memories are right...

**Comment by tiwillia** [comment]
@deads2k I've been running it locally. Almost 2 hours in and its still going, [Slow] was an understatement.

I'm putting together the `test/extended/slow.sh` as you requested (which is how I'm running the above).

**Comment by deads2k** [comment]
> @deads2k I've been running it locally. Almost 2 hours in and its still going, [Slow] was an understatement.

Something might be wrong.  Just that test should have been 7 minutes according to the comment.

**Comment by tiwillia** [comment]
Yup, I apologize, running into some issues running them locally.

Seeing it fail now due to:
```
[...]
Mar 11 13:14:21.638: INFO: Error getting replication controllers in namespace 'kube-system': Get https://172.18.9.218:8443/api/v1/namespaces/kube-system/replicationcontrollers: x509: certificate signed by unknown authority
Mar 11 13:14:23.633: INFO: Error getting replication controllers in namespace 'kube-system': Get https://172.18.9.218:8443/api/v1/namespaces/kube-system/replicationcontrollers: x509: certificate signed by unknown authority
Mar 11 13:14:25.633: INFO: Error getting replication controllers in namespace 'kube-system': Get https://172.18.9.218:8443/api/v1/namespaces/kube-system/replicationcontrollers: x509: certificate signed by unknown authority
Mar 11 13:14:25.635: INFO: Error getting replication controllers in namespace 'kube-system': Get https://172.18.9.218:8443/api/v1/namespaces/kube-system/replicationcontrollers: x509: certificate signed by unknown authority
Mar 11 13:14:25.635: FAIL: Error waiting for all pods to be running and ready: Not all pods in namespace 'kube-system' running and ready within 10m0s  
```

Not really getting to the actual test, trying to figure out what I've done incorrectly with the test setup.

**Comment by deads2k** [comment]
> Not really getting to the actual test, trying to figure out what I've done incorrectly with the test setup.

@jayunit100 any advice?

**Comment by deads2k** [comment]
```
  Mar 11 13:32:01.170: Only supported for providers [gce] (not )
```

Didn't see that coming.  It's not obvious to me why.  See about eliminating the restriction and figuring out what happens.

**Comment by jayunit100** [comment]
If the upstream test only runs for GCE but other providers support it, we should definetly modify upstream to get rid of the skip.

**Comment by tiwillia** [comment]
Commented out the gce provider restriction in the garbage_collector test, trying again now.

**Comment by deads2k** [comment]
Don't know whether the failure is legit, but it failed.  I don't know enough about the component to write a test quickly, but we will need a good test before merge.

**Comment by jayunit100** [comment]
Iirc Garbage collector e2es are pretty complex, not even sure if they are stable,... Esp on a one node system like the one we run for ci.  

I think fine grained manual verification is gonna be needed for this.

**Comment by deads2k** [comment]
> I think fine grained manual verification is gonna be needed for this.

@tiwillia if you have to go with that approach, make sure that its noted on the card, carefully review the test, and see if there's a way to indicate that it'll need to be run in a regression bucket because its the only test we'll have for it.

**Comment by tiwillia** [comment]
I believe the issue with the test is that the terminated pod gc threshold option is never set on the controller manager in the testing suite. The pod gc test expects this to be set to 100. The default is 12500, so the test is failing. 

Modified the default to verify. Will need to figure out how to set controller manager options in the test if I'm correct.

**Comment by tiwillia** [comment]
@soltysh @mfojtik are you aware of any reasonable way to set a kube controller manager configuration option from an extended test? Is there a configuration file maybe where the kube controller manager options can be changed, like we do in `test/extended/alternate_certs.sh`?

**Comment by deads2k** [comment]
Looks like its plumbed in via the `--terminated-pod-gc-threshold` flag, so I'd expect it to be `master-config.yaml|kubernetesMasterConfig.controllerArguments[terminated-pod-gc-threshold] = 100`.

If you can make the test run in less than 30 seconds or so, you could also create an integration test that allows you manipulate the config in objects.

**Comment by tiwillia** [comment]
I don't think we'll be able to make a test that runs in less than 30 seconds or so. The gc check period is set to a constant of 20 seconds, so we'd wait at least that long after creating pods for the test to ensure that the garbage collector has had time to reap pods.

This is going to have to be something a little unorthodox I think. I've mimiced the admittedly hacky was the `test/extended/alternate_certs.sh` test modifies the `master-config.yaml` to set the `TerminatedPodGCThreshold` properly for the test.

Also added the `test/extended/slow.sh`, which for now just runs the garbage collector test (and is where the sed command is used).

I assume this might work now? [test][extended:slow]

**Comment by smarterclayton** [comment]
You could set an aggressive garbage collection interval in the default
extended setup, and then add your test there.

**Comment by tiwillia** [comment]
@smarterclayton are you saying I could modify the `gcCheckPeriod` from a test? Not sure how I would modify this constant:
https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/gc/gc_controller.go#L40

**Comment by smarterclayton** [comment]
I was referring to setting TerminatedPodGCThreshold to what you need
in core.sh, so that you didn't have to create a new launch profile.

**Comment by tiwillia** [comment]
@smarterclayton I could do that. @deads2k requested I create a new launch profile for slow tests a few comments up. If you'd prefer putting everything in `core.sh`, I'm happy to delete the `test/extended/slow.sh` and put the config modification into `test/extended/core.sh`

**Comment by smarterclayton** [comment]
I think slow.sh should really just be core.sh with a different focus.
Everything slow should be marked slow.

**Comment by smarterclayton** [comment]
But even then slow.sh would be a wrapper to core.sh

**Comment by tiwillia** [comment]
re[test] 'ing to make sure the sed is correct, test failed in 5 mins during build. I'll probably end up moving it to core.sh and elminating slow.sh.

**Comment by tiwillia** [comment]
Tests should be passing now. UPSTREAM commit submitted and marked properly here. Not sure how to remove the 'slow' extended test from jenkins.

**Comment by tiwillia** [comment]
It _looks_ like this is failing because the `test/extended/slow.sh` script no longer exists:
```
bash: line 9: test/extended/slow.sh: No such file or directory
```

How would I go about removing a script from testing? @soltysh @deads2k

**Comment by deads2k** [comment]
[test] [extended:core(collector)]

**Comment by deads2k** [comment]
it was still in one of your comments in this pull.  I took the liberty of removing it.

**Comment by deads2k** [comment]
[test]

**Comment by tiwillia** [comment]
Ever since the rebase, in my testing, I get the following error:
```
error reading config: [pos 2131]: only encoded map or array can be decoded into a slice (0)
```

With the following in the `master-config.yaml`:
```
kubernetesMasterConfig:
  [...]
  controllerArguments: 
    terminated-pod-gc-threshold: 100
```

`controllerArguments` still look to be defined as key value pairs. I'd like to see if the automated testing fails with the same error.

**Comment by liggitt** [comment]
controllerArguments are key to array of values:

```
controllerArguments:
  terminated-pod-gc-threshold: [100]
```

File: test/extended/core.sh [hunk 3]
```diff
@@ -79,6 +79,10 @@ if [[ -z ${TEST_ONLY+x} ]]; then
   fi
   echo "[INFO] Using VOLUME_DIR=${VOLUME_DIR}"
 
+  # This is a bit hacky, but set the pod gc threshold appropriately for the garbage_collector test.
+  os::util::sed 's/\(controllerArguments:\ \)null/\1\n\ \ \ \ terminated-pod-gc-threshold: 100/' \
```
Comment (liggitt) [inline test/extended/core.sh]: shouldn't need to escape spaces, right?
```
os::util::sed 's/^  controllerArguments: null/  controllerArguments: \n    terminated-pod-gc-threshold: ["100"]/'
```

**Comment by tiwillia** [comment]
Thank you @liggitt 

The '100' had to be put in double quotes too:
```
  controllerArguments:
    terminated-pod-gc-threshold: ["100"]
```

**Comment by tiwillia** [comment]
re[test]

**Comment by tiwillia** [comment]
re[test] please

**Comment by tiwillia** [comment]
The jenkins job appears to still be running the slow.sh test script. It looks to be passing otherwise.

@deads2k any idea why it continues to try to use the slow test script still?

**Comment by soltysh** [comment]
@tiwillia I'd suggest opening a new PR, apparently somewhere in the comments there's slow defined and jenkins is picking that up. Open the new PR - that will be the fastest way to handle that problem.

**Comment by tiwillia** [comment]
Closing as per @soltysh 's recommendation

**Pull request closed without merging**